File: skills/agent-creator/scripts/grade_agent_eval.py
#!/usr/bin/env python3
"""Grade paired custom-agent eval outputs with deterministic or LLM grading."""

from __future__ import annotations

import argparse
import json
import os
import re
import shlex
import subprocess
from pathlib import Path


def deterministic_grade(assertion: str, response: str) -> tuple[bool, str] | None:
    match = re.fullmatch(r"contains:\s*(.+)", assertion, re.IGNORECASE)
    if match:
        needle = match.group(1)
        passed = needle.casefold() in response.casefold()
        return passed, f"Expected response to contain: {needle!r}"
    match = re.fullmatch(r"not-contains:\s*(.+)", assertion, re.IGNORECASE)
    if match:
        needle = match.group(1)
        passed = needle.casefold() not in response.casefold()
        return passed, f"Expected response not to contain: {needle!r}"
    match = re.fullmatch(r"regex:\s*(.+)", assertion, re.IGNORECASE)
    if match:
        pattern = match.group(1)
        try:
            passed = re.search(pattern, response, re.MULTILINE) is not None
        except re.error as error:
            raise ValueError(f"Invalid assertion regex {pattern!r}: {error}") from error
        return passed, f"Regex {pattern!r} {'matched' if passed else 'did not match'}"
    return None


def llm_grade(prompt: str, response: str, assertion: str, command: tuple[str, ...], model: str | None) -> dict:
    grading_prompt = f"""Grade one custom-agent evaluation assertion.

Task:
{prompt}

Agent response:
{response}

Assertion:
{assertion}

Return JSON only with exactly these fields:
{{"passed": true, "evidence": "specific evidence from the response"}}
Use passed=false when the evidence is missing, contradicted, or unverifiable.
"""
    args = [*command, "run", "--no-session", "--quiet", "--output-format", "text", "--instructions", "-"]
    if model:
        args.extend(["--model", model])
    result = subprocess.run(args, input=grading_prompt, capture_output=True, text=True, timeout=300)
    if result.returncode != 0:
        raise RuntimeError(f"Grader failed: {result.stderr.strip()}")
    text = result.stdout.strip()
    match = re.search(r"\{.*\}", text, re.DOTALL)
    if not match:
        raise RuntimeError(f"Grader did not return JSON: {text}")
    document = json.loads(match.group(0))
    if not isinstance(document.get("passed"), bool) or not isinstance(document.get("evidence"), str):
        raise RuntimeError(f"Invalid grader result: {document}")
    return document


def grade_run(
    run_dir: Path,
    prompt: str,
    assertions: list[str],
    command: tuple[str, ...],
    model: str | None,
    use_llm: bool,
):
    response_path = run_dir / "outputs" / "response.md"
    response = response_path.read_text(encoding="utf-8") if response_path.is_file() else ""
    results = []
    for assertion in assertions:
        deterministic = deterministic_grade(assertion, response)
        if deterministic is not None:
            passed, evidence = deterministic
            result = {"passed": passed, "evidence": evidence}
        elif use_llm:
            result = llm_grade(prompt, response, assertion, command, model)
        else:
            result = {
                "passed": False,
                "evidence": "Assertion requires LLM grading; rerun with --llm-grader",
            }
        results.append({"text": assertion, **result})

    passed = sum(1 for result in results if result["passed"])
    total = len(results)
    timing = {}
    timing_path = run_dir / "timing.json"
    if timing_path.is_file():
        timing = json.loads(timing_path.read_text(encoding="utf-8"))
    grading = {
        "expectations": results,
        "summary": {
            "passed": passed,
            "failed": total - passed,
            "total": total,
            "pass_rate": passed / total if total else 0.0,
        },
        "timing": timing,
    }
    (run_dir / "grading.json").write_text(json.dumps(grading, indent=2) + "\n")


def main():
    parser = argparse.ArgumentParser(description="Grade custom-agent eval outputs")
    parser.add_argument("workspace")
    parser.add_argument("--llm-grader", action="store_true", help="Use Goose for free-form assertions")
    parser.add_argument("--goose-cli", default=os.environ.get("AGENT_CREATOR_GOOSE_CLI", "goose"))
    parser.add_argument("--model", default=None)
    args = parser.parse_args()

    workspace = Path(args.workspace).expanduser().resolve()
    command = tuple(shlex.split(args.goose_cli))
    graded = 0
    for eval_dir in sorted(workspace.glob("eval-*")):
        metadata_path = eval_dir / "eval_metadata.json"
        if not metadata_path.is_file():
            continue
        metadata = json.loads(metadata_path.read_text(encoding="utf-8"))
        for run_dir in sorted(path for path in eval_dir.iterdir() if path.is_dir()):
            if not (run_dir / "outputs").is_dir():
                continue
            grade_run(
                run_dir,
                metadata.get("prompt", ""),
                metadata.get("assertions", []),
                command,
                args.model,
                args.llm_grader,
            )
            graded += 1
    print(f"Graded {graded} runs in {workspace}")


if __name__ == "__main__":
    main()

File: skills/agent-creator/scripts/grade_agent_eval.ts
#!/usr/bin/env node
// Grade paired custom-agent eval outputs with deterministic or LLM grading.
import { readFileSync, writeFileSync, existsSync, readdirSync, statSync } from "node:fs";
import { join, resolve } from "node:path";
import { parseArgs } from "node:util";
import { execFile } from "node:child_process";

function execFileWithInput(
  bin: string,
  args: string[],
  input: string,
  timeoutMs: number
): Promise<{ stdout: string; stderr: string }> {
  return new Promise((resolvePromise, reject) => {
    const child = execFile(
      bin,
      args,
      { timeout: timeoutMs, maxBuffer: 1024 * 1024 * 16, encoding: "utf-8" },
      (error, stdout, stderr) => {
        if (error) {
          (error as any).stderr = stderr;
          reject(error);
          return;
        }
        resolvePromise({ stdout, stderr });
      }
    );
    child.stdin?.write(input);
    child.stdin?.end();
  });
}

interface GradeResult {
  passed: boolean;
  evidence: string;
}

export function deterministicGrade(assertion: string, response: string): GradeResult | null {
  let match = /^contains:\s*(.+)$/is.exec(assertion.trim());
  if (match) {
    const needle = match[1];
    const passed = response.toLowerCase().includes(needle.toLowerCase());
    return { passed, evidence: `Expected response to contain: '${needle}'` };
  }
  match = /^not-contains:\s*(.+)$/is.exec(assertion.trim());
  if (match) {
    const needle = match[1];
    const passed = !response.toLowerCase().includes(needle.toLowerCase());
    return { passed, evidence: `Expected response not to contain: '${needle}'` };
  }
  match = /^regex:\s*(.+)$/is.exec(assertion.trim());
  if (match) {
    const pattern = match[1];
    let passed: boolean;
    try {
      passed = new RegExp(pattern, "m").test(response);
    } catch (error) {
      throw new Error(`Invalid assertion regex '${pattern}': ${(error as Error).message}`);
    }
    return { passed, evidence: `Regex '${pattern}' ${passed ? "matched" : "did not match"}` };
  }
  return null;
}

async function llmGrade(
  prompt: string,
  response: string,
  assertion: string,
  command: string[],
  model: string | undefined
): Promise<{ passed: boolean; evidence: string }> {
  const gradingPrompt = `Grade one custom-agent evaluation assertion.

Task:
${prompt}

Agent response:
${response}

Assertion:
${assertion}

Return JSON only with exactly these fields:
{"passed": true, "evidence": "specific evidence from the response"}
Use passed=false when the evidence is missing, contradicted, or unverifiable.
`;
  const args = ["run", "--no-session", "--quiet", "--output-format", "text", "--instructions", "-"];
  if (model) args.push("--model", model);

  const [bin, ...baseArgs] = command;
  let stdout: string;
  try {
    const result = await execFileWithInput(bin, [...baseArgs, ...args], gradingPrompt, 300_000);
    stdout = result.stdout;
  } catch (error: any) {
    throw new Error(`Grader failed: ${(error.stderr ?? error.message ?? "").trim()}`);
  }
  const text = stdout.trim();
  const match = /\{[\s\S]*\}/.exec(text);
  if (!match) throw new Error(`Grader did not return JSON: ${text}`);
  const document = JSON.parse(match[0]);
  if (typeof document.passed !== "boolean" || typeof document.evidence !== "string") {
    throw new Error(`Invalid grader result: ${JSON.stringify(document)}`);
  }
  return document;
}

async function gradeRun(
  runDir: string,
  prompt: string,
  assertions: string[],
  command: string[],
  model: string | undefined,
  useLlm: boolean
) {
  const responsePath = join(runDir, "outputs", "response.md");
  const response = existsSync(responsePath) ? readFileSync(responsePath, "utf-8") : "";
  const results: Array<{ text: string } & GradeResult> = [];
  for (const assertion of assertions) {
    const deterministic = deterministicGrade(assertion, response);
    let result: GradeResult;
    if (deterministic !== null) {
      result = deterministic;
    } else if (useLlm) {
      result = await llmGrade(prompt, response, assertion, command, model);
    } else {
      result = {
        passed: false,
        evidence: "Assertion requires LLM grading; rerun with --llm-grader",
      };
    }
    results.push({ text: assertion, ...result });
  }

  const passed = results.filter((r) => r.passed).length;
  const total = results.length;
  let timing: unknown = {};
  const timingPath = join(runDir, "timing.json");
  if (existsSync(timingPath)) {
    timing = JSON.parse(readFileSync(timingPath, "utf-8"));
  }
  const grading = {
    expectations: results,
    summary: {
      passed,
      failed: total - passed,
      total,
      pass_rate: total ? passed / total : 0.0,
    },
    timing,
  };
  writeFileSync(join(runDir, "grading.json"), `${JSON.stringify(grading, null, 2)}\n`);
}

function isDirectory(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}

async function main() {
  const { positionals, values } = parseArgs({
    args: process.argv.slice(2),
    allowPositionals: true,
    options: {
      "llm-grader": { type: "boolean", default: false },
      "goose-cli": { type: "string", default: process.env.AGENT_CREATOR_GOOSE_CLI ?? "goose" },
      model: { type: "string" },
    },
  });

  const [workspaceArg] = positionals;
  if (!workspaceArg) {
    console.error("usage: grade_agent_eval.js <workspace> [--llm-grader] [--goose-cli <cmd>] [--model <name>]");
    process.exit(2);
  }

  const workspace = resolve(workspaceArg);
  const command = (values["goose-cli"] as string).split(/\s+/).filter(Boolean);
  let graded = 0;

  const evalDirs = readdirSync(workspace)
    .filter((name) => name.startsWith("eval-"))
    .sort();
  for (const evalName of evalDirs) {
    const evalDir = join(workspace, evalName);
    const metadataPath = join(evalDir, "eval_metadata.json");
    if (!existsSync(metadataPath)) continue;
    const metadata = JSON.parse(readFileSync(metadataPath, "utf-8"));
    const runDirs = readdirSync(evalDir)
      .filter((name) => isDirectory(join(evalDir, name)))
      .sort();
    for (const runName of runDirs) {
      const runDir = join(evalDir, runName);
      if (!isDirectory(join(runDir, "outputs"))) continue;
      await gradeRun(
        runDir,
        metadata.prompt ?? "",
        metadata.assertions ?? [],
        command,
        values.model as string | undefined,
        values["llm-grader"] as boolean
      );
      graded += 1;
    }
  }
  console.log(`Graded ${graded} runs in ${workspace}`);
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((error) => {
    console.error(error?.message ?? error);
    process.exit(1);
  });
}
